osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiplayerSpectatorLeaderboard.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Timing;
+using osu.Game.Online.Spectator;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Screens.Play.HUD;
+
+namespace osu.Game.Screens.OnlinePlay.Multiplayer.Spectate
+{
+    public class MultiplayerSpectatorLeaderboard : MultiplayerGameplayLeaderboard
+    {
+        private readonly Dictionary<int, TrackedUserData> trackedData = new Dictionary<int, TrackedUserData>();
+
+        public MultiplayerSpectatorLeaderboard(ScoreProcessor scoreProcessor, int[] userIds)
+            : base(scoreProcessor, userIds)
+        {
+        }
+
+        public void AddClock(int userId, IClock source) => trackedData[userId] = new TrackedUserData(source);
+
+        public void RemoveClock(int userId) => trackedData.Remove(userId);
+
+        protected override void OnIncomingFrames(int userId, FrameDataBundle bundle)
+        {
+            if (!trackedData.TryGetValue(userId, out var data))
+                return;
+
+            data.Frames.Add(new TimedFrameHeader(bundle.Frames.First().Time, bundle.Header));
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            foreach (var (userId, data) in trackedData)
+            {
+                var targetTime = data.Clock.CurrentTime;
+
+                if (data.Frames.Count == 0)
+                    continue;
+
+                int frameIndex = data.Frames.BinarySearch(new TimedFrameHeader(targetTime));
+                if (frameIndex < 0)
+                    frameIndex = ~frameIndex;
+                frameIndex = Math.Clamp(frameIndex - 1, 0, data.Frames.Count - 1);
+
+                SetCurrentFrame(userId, data.Frames[frameIndex].Header);
+            }
+        }
+
+        private class TrackedUserData
```

Comment:
Not sure how I feel about having this class in two places with the same name. On top of that, the base implementation seems to be getting overridden but is still existing in a weird unused state. Did you consider making a `CreateUserData` function and having this implementation create a derived version with the extra required data? Or simpler, just add the time/frame data to the base implementation and call it a day (probably fine for now?)

Reply:
I've reworked it with a `CreateUserData` method as you suggested.